lib/core/include/irods/fully_qualified_username.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef IRODS_FULLY_QUALIFIED_USERNAME_HPP
+#define IRODS_FULLY_QUALIFIED_USERNAME_HPP
+
+/// \file
+
+#include "irods/irods_exception.hpp"
+#include "irods/rodsErrorTable.h"
+
+#include <string_view>
+
+namespace irods::experimental
+{
+    /// A class which is designed to enforce use of fully-qualified iRODS usernames.
+    ///
+    /// This class should be used as a building block for other libraries and utilities.
+    ///
+    /// Instances of this class are readonly.
+    ///
+    /// \since 4.3.1
+    class fully_qualified_username
+    {
+      public:
+        /// Constructs a fully_qualified_username.
+        ///
+        /// \param[in] _name The part of the username preceding the pound sign.
+        /// \param[in] _zone The part of the username following the pound sign.
+        ///
+        /// \throws irods::exception If an error occurs.
```

Comment:
I've started moving away from mentioning that because it's difficult to maintain and keep true.

I personally think letting devs know that a particular type of exception can be throw is better. With that, they know they have to catch _N_ types of exceptions. From there, they can interrogate the exceptions for what caused it.

Thoughts?

Reply:
I think that's fine.

Just thought it could be helpful to mention that `_name` and `_zone` are required to be non-empty, whether that's in this part or in the input parameter descriptions. No need to do anything here.

Reply:
Ah. That makes sense.

Will update the parameter descriptions.

Reply:
Hmm, then again the name of the type says it all. This is hard lol.

Will think about this more... again :-)

Reply:
Will update the documentation to include more words. It's just never a bad idea to be explicit. In this particular case, it's obvious what to write.